commons/src/main/java/com/github/phantazmnetwork/commons/component/ReflectiveComponentBuilder.java

```diff
@@ -0,0 +1,263 @@
+    private final KeyedFactoryRegistry factoryRegistry;
+
+    /**
+     * Creates a new instance of this class which will use the provided registries to handle component creation.
+     *
+     * @param configRegistry  the {@link KeyedConfigRegistry} used to register configuration processors
+     * @param factoryRegistry the {@link KeyedFactoryRegistry} used to register component factories
+     */
+    public ReflectiveComponentBuilder(@NotNull KeyedConfigRegistry configRegistry,
+                                      @NotNull KeyedFactoryRegistry factoryRegistry) {
+        this.configRegistry = Objects.requireNonNull(configRegistry, "configRegistry");
+        this.factoryRegistry = Objects.requireNonNull(factoryRegistry, "factoryRegistry");
+    }
+
+    @Override
+    public void registerComponentClass(@NotNull Class<?> component) throws ComponentException {
+        ComponentModel annotation = component.getAnnotation(ComponentModel.class);
+        if (annotation == null) {
+            throw new ComponentException(
+                    "Class " + component.getTypeName() + " must have the ComponentModel annotation");
+        }
+
+        String value = annotation.value();
+
+        Key componentKey = getKey(component, value);
+        boolean hasProcessor = true;
+        if (!configRegistry.hasProcessor(componentKey)) {
+            KeyedConfigProcessor<?> processor = getProcessor(component);
+
+            if (processor == null) {
+                hasProcessor = false;
+            }
+            else {
+                configRegistry.registerProcessor(componentKey, processor);
+            }
+        }
+
+        if (!factoryRegistry.hasFactory(componentKey)) {
+            KeyedFactory<?, ?> factory = getFactory(component, hasProcessor);
+            factoryRegistry.registerFactory(componentKey, factory);
+        }
+    }
+
+    @Override
+    public <TData extends Keyed, TComponent> TComponent makeComponent(@NotNull ConfigNode node,
+                                                                      @NotNull DependencyProvider provider)
+            throws ComponentException {
+        Key dataKey;
+        Keyed data = null;
+        try {
+            dataKey = configRegistry.extractKey(node);
+
+            if (configRegistry.hasProcessor(dataKey)) {
+                data = configRegistry.deserialize(node);
+            }
+        }
+        catch (ConfigProcessException e) {
+            throw new ComponentException(e);
+        }
+
+        KeyedFactory<TData, ?> factory = factoryRegistry.getFactory(dataKey);
+        if (factory == null) {
+            throw new ComponentException("Unable to find a suitable factory for " + dataKey);
+        }
+
+        if (!provider.load(factory.dependencies())) {
+            throw new ComponentException("Unable to prepare dependencies for data " + dataKey);
+        }
+
+        //noinspection unchecked
+        return (TComponent)factory.make(provider, (TData)data);
```

Comment:
I think it'd be better to get rid of the TData generic so that it's not part of the superinterface's method signature without it being used since the generic param is kind of an implementation detail, so create another private method with a generic to make sure it matches up

Reply:
just messed around with it and it turns out that TData isn't even needed by the subclass